### ISSUE TITLE
Inaccurate key type during dict and list comprehension with string literal keys

### ISSUE THREAD

**Issue by MysticMia** [microsoft/pyright#10835]
**Describe the bug**
The type of a dictionary with string literal keys changes when using the keys in dictionary comprehension

**Code or Screenshots**
```py
from typing import Literal
type Foo = Literal["foo"]

bar1: dict[Foo, int] = {'foo': 1}
bar2: dict[Foo, int] = {k: v for k, v in bar1.items()}

baz1: dict[Foo, int] = bar1 | bar2
baz2: dict[Foo, int] = bar1 | {k: v for k, v in bar1.items()}  # error
```
```
Type "dict[str, int]" is not assignable to declared type "dict[Foo, int]"
  "dict[str, int]" is not assignable to "dict[Foo, int]"
    Type parameter "_KT@dict" is invariant, but "str" is not the same as "Foo" (reportAssignmentType)
```

When removing the type annotation for `bar2`, the same error also occurs for `baz1`:
```py
from typing import Literal
type Foo = Literal["foo"]

bar1: dict[Foo, int] = {'foo': 1}
bar2 = {k: v for k, v in bar1.items()}  # note the removed type annotation
# PyCharm types it as dict[str, int], but its interpreted type is dict[Foo, int] as expected.

baz1: dict[Foo, int] = bar1 | bar2  # error xyz
baz2: dict[Foo, int] = bar1 | {k: v for k, v in bar1.items()}  # error xyz (full error is the same as above)
```

<img width="902" height="282" alt="Image" src="https://github.com/user-attachments/assets/8808ada3-2357-4ab4-a74d-bd5dc5ba32a5" />

**VS Code extension or command-line**
This bug occurs when using `pyright 1.1.404` in the command line, and when running it in JetBrains PyCharm using the Pyright plugin (version 0.12.0).

**Comment by MysticMia** [comment]
The same happens for list comprehension:

<img width="810" height="297" alt="Image" src="https://github.com/user-attachments/assets/1233cc1b-f812-41af-99aa-a9d4c259ac9d" />

Meanwhile, giving `bar3` a type annotation resolves this issue:

<img width="379" height="147" alt="Image" src="https://github.com/user-attachments/assets/caf37c7f-2cc1-430e-882a-b48e10a75cdd" />

```py
from typing import Literal

type Foo = Literal["foo"]

bar1: dict[Foo, int] = {'foo': 1}
bar3: list[Foo] = [k for k in bar1]

baz0: list[Foo] = bar3 + list(bar1)
```

**Comment by erictraut** [comment]
This is expected behavior, so I don't consider it a bug.

By default, when pyright infers the type of dict, list or set expressions, it does not retain literals. It widens them to their non-literal forms. This default behavior can be overridden through the use of [bidirectional type inference](https://microsoft.github.io/pyright/#/type-inference?id=bidirectional-type-inference-expected-types). That explains why `bar2: dict[Foo, int] = {k: v for k, v in bar1.items()}` works. 

Bidirectional type inference cannot be used when inferring the types of the LHS or RHS of a binary operator like `+` or `|` because those are complex operators that involve calls to (potentially) multiple magic methods. For example, `+` involves calls to either `__add__` or `__radd__`. To determine which call to use, pyright must evaluate the LHS and RHS without the benefit of bidirectional type inference.

If you want this to type check without errors, you'll need to use a form that enables bidirectional type inference. You can do this by assigning to (annotated) intermediate variables like in your code sample. You can also wrap in a constructor call with explicit specialization like this: `baz2: dict[Foo, int] = bar1 | dict[Foo, int]({k: v for k, v in bar1.items()})`.
